src/smt/theory_str_noodler/theory_str_noodler.cpp

```diff
@@ -1889,6 +1893,101 @@ namespace smt::noodler {
         }
     }
 
+    /**
+     * @brief Handle str.<=
+     * Translates to the following axiom
+     * 
+     * x <= y -> x = y | x < y
+     * @param e str.<= predicate
+     */
+    void theory_str_noodler::handle_lex_leq(expr *e) {
+        if(axiomatized_persist_terms.contains(e))
+            return;
+
+        axiomatized_persist_terms.insert(e);
+        STRACE("str", tout  << "handle str.<= " << mk_pp(e, m) << std::endl;);
+
+        expr *x = nullptr, *y = nullptr;
+        VERIFY(m_util_s.str.is_le(e, x, y));
+      
+        expr_ref e_lt(m_util_s.str.mk_lex_lt(x, y), m);
+        expr_ref x_y(m.mk_eq(x,y), m);
+        literal lit_e_lt = mk_literal(e_lt);
+        literal lit_e = mk_literal(e);
+        literal lit_x_y = mk_literal(x_y);
+        // x <= y -> x = y | x < y
+        add_axiom({~lit_e, lit_e_lt, lit_x_y});
+    }
+
+    /**
+     * @brief Handle str.<
+     * Translates to the following theory axioms.
+     * 
+     * x = eps & y != eps -> x < y
+     * x < y & x != eps -> x = u.v1.w1
```

Comment:
As @Adda0 said, `x < y & x != eps -> y = u.v2.w2` is missing here. But then also for `y` we must have `y != eps`, and you also ignore the cases `x != eps & y != eps` and `x = eps & y = eps`. I think it does not work properly anyway, it does not handle `~(x<y)` correctly I think.
I would probably do it this way:
`x < y -> (y != eps & (x = eps | (x = u.v1.w1 & y = u.v2.w2 & v1,v2 in re.allchar & to_code(v1) < to_code(v2))))`
`!(x < y) -> (x = y | y < x)`

Reply:
if `y = eps` str.< is false. I think we do not need to mention it inside axioms.

Reply:
The negation of < is indeed missing. It should be there I think.

Reply:
I added the axiom `!(x < y) -> (x = y | y < x)`. How did you want to modify `x = eps & y != eps -> x < y`?

Reply:
I would put `x < y & x = eps -> y != eps` so it is obvious it is case split on `x = eps`.